panic when applying a saved plan file created with "terraform plan -destroy -out=tfplan"

### Terraform Version

```
Terraform v0.12.0-alpha1
```

### Crash Output
```
aws_subnet.all[1]: Destroying... [id=subnet-050215c1492df555c]
aws_subnet.all[2]: Destroying... [id=subnet-0079c70f5e936a9a9]
aws_subnet.all[0]: Destroying... [id=subnet-0fe6fb156f4d13209]
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x7248f2]

goroutine 78 [running]:
github.com/zclconf/go-cty/cty.Type.Equals(0x0, 0x0, 0x37855e0, 0xc0002d00d8, 0x200)
	/opt/gopath/src/github.com/hashicorp/terraform/vendor/github.com/zclconf/go-cty/cty/type.go:39 +0x22
github.com/zclconf/go-cty/cty.testConformance(0x0, 0x0, 0x37855e0, 0xc0002d00d8, 0xc000868630, 0x0, 0x0, 0xc000868630)
	/opt/gopath/src/github.com/hashicorp/terraform/vendor/github.com/zclconf/go-cty/cty/type_conform.go:37 +0xae
github.com/zclconf/go-cty/cty.Type.TestConformance(0x0, 0x0, 0x37855e0, 0xc0002d00d8, 0x37855e0, 0xc0002d00d8, 0xc0008686b8)
	/opt/gopath/src/github.com/hashicorp/terraform/vendor/github.com/zclconf/go-cty/cty/type_conform.go:27 +0x7a
github.com/zclconf/go-cty/cty/msgpack.Marshal(0x0, 0x0, 0x0, 0x0, 0x37855e0, 0xc0002d00d8, 0xc0005c2000, 0x144, 0x193, 0x0, ...)
	/opt/gopath/src/github.com/hashicorp/terraform/vendor/github.com/zclconf/go-cty/cty/msgpack/marshal.go:19 +0x67
github.com/hashicorp/terraform/plugin.(*GRPCProvider).ApplyResourceChange(0xc0002dc900, 0xc0002bd9f0, 0xa, 0x37855e0, 0xc0002d00b0, 0x2bfee80, 0xc0005e5d70, 0x0, 0x0, 0x0, ...)
	/opt/gopath/src/github.com/hashicorp/terraform/plugin/grpc_provider.go:410 +0x247
github.com/hashicorp/terraform/terraform.(*EvalApply).Eval(0xc000a04410, 0x3798600, 0xc00001ef70, 0x2, 0x2, 0x43ef89, 0xc0000d40c0)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval_apply.go:71 +0x61b
github.com/hashicorp/terraform/terraform.EvalRaw(0x376d7e0, 0xc000a04410, 0x3798600, 0xc00001ef70, 0x0, 0x0, 0x0, 0x27)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval.go:57 +0x167
github.com/hashicorp/terraform/terraform.(*EvalIf).Eval(0xc0005e5080, 0x3798600, 0xc00001ef70, 0x2, 0x2, 0x0, 0x0)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval_if.go:21 +0x9c
github.com/hashicorp/terraform/terraform.EvalRaw(0x376da20, 0xc0005e5080, 0x3798600, 0xc00001ef70, 0x0, 0x0, 0x0, 0x0)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval.go:57 +0x167
github.com/hashicorp/terraform/terraform.(*EvalSequence).Eval(0xc0002c8b00, 0x3798600, 0xc00001ef70, 0x2, 0x2, 0x42, 0x0)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval_sequence.go:20 +0xfd
github.com/hashicorp/terraform/terraform.EvalRaw(0x376dca0, 0xc0002c8b00, 0x3798600, 0xc00001ef70, 0x2d, 0x0, 0x0, 0x2d)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval.go:57 +0x167
github.com/hashicorp/terraform/terraform.(*EvalOpFilter).Eval(0xc0005e50e0, 0x3798600, 0xc00001ef70, 0x2, 0x2, 0x6, 0x10)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval_filter_operation.go:37 +0x4c
github.com/hashicorp/terraform/terraform.EvalRaw(0x376db60, 0xc0005e50e0, 0x3798600, 0xc00001ef70, 0x2bef760, 0x5c23181, 0x28dfb80, 0xc000354510)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval.go:57 +0x167
github.com/hashicorp/terraform/terraform.Eval(0x376db60, 0xc0005e50e0, 0x3798600, 0xc00001ef70, 0xc0005e50e0, 0x376db60, 0xc0005e50e0, 0x0)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval.go:35 +0x4d
github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x332e3a0, 0xc0002c9300, 0x0, 0x0, 0x0)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/graph.go:90 +0x10d1
github.com/hashicorp/terraform/dag.(*Walker).walkVertex(0xc00013a300, 0x332e3a0, 0xc0002c9300, 0xc00008a180)
	/opt/gopath/src/github.com/hashicorp/terraform/dag/walk.go:392 +0x373
created by github.com/hashicorp/terraform/dag.(*Walker).Update
	/opt/gopath/src/github.com/hashicorp/terraform/dag/walk.go:314 +0x9d2
```

### Expected Behavior
The AWS provider should have deleted the instances as planned.

### Actual Behavior
The provider (most likely: the shared provider SDK) seems to have triggered a panic during the decoding of the provider request to apply the change.

### Steps to Reproduce

With resources already existing from a previous apply:

1. `terraform plan -destroy -out=tfplan`
2. `terraform apply tfplan`

--- a/plans/changes_src.go
+++ b/plans/changes_src.go
@@ -166,13 +166,21 @@ type ChangeSrc struct {
 // to call the corresponding Decode method of that struct rather than working
 // directly with its embedded Change.
 func (cs *ChangeSrc) Decode(ty cty.Type) (*Change, error) {
-	before, err := cs.Before.Decode(ty)
-	if err != nil {
-		return nil, fmt.Errorf("error decoding 'before' value: %s", err)
+	var err error
+	before := cty.NullVal(ty)
+	after := cty.NullVal(ty)
+
+	if len(cs.Before) > 0 {
+		before, err = cs.Before.Decode(ty)
+		if err != nil {
+			return nil, fmt.Errorf("error decoding 'before' value: %s", err)
+		}
 	}
-	after, err := cs.After.Decode(ty)
-	if err != nil {
-		return nil, fmt.Errorf("error decoding 'after' value: %s", err)
+	if len(cs.After) > 0 {
+		after, err = cs.After.Decode(ty)
+		if err != nil {
+			return nil, fmt.Errorf("error decoding 'after' value: %s", err)
+		}
 	}
 	return &Change{
 		Action: cs.Action,

--- a/plans/dynamic_value.go
+++ b/plans/dynamic_value.go
@@ -86,7 +86,7 @@ func (v DynamicValue) ImpliedType() (cty.Type, error) {
 
 // Copy produces a copy of the receiver with a distinct backing array.
 func (v DynamicValue) Copy() DynamicValue {
-	if len(v) == 0 {
+	if v == nil {
 		return nil
 	}
 

--- a/plans/planfile/tfplan_test.go
+++ b/plans/planfile/tfplan_test.go
@@ -26,7 +26,7 @@ func TestTFPlanRoundTrip(t *testing.T) {
 					Addr: addrs.OutputValue{Name: "bar"}.Absolute(addrs.RootModuleInstance),
 					ChangeSrc: plans.ChangeSrc{
 						Action: plans.Create,
-						After:  mustNewDynamicValueStr("bar value"),
+						After:  mustDynamicOutputValue("bar value"),
 					},
 					Sensitive: false,
 				},
@@ -34,8 +34,8 @@ func TestTFPlanRoundTrip(t *testing.T) {
 					Addr: addrs.OutputValue{Name: "baz"}.Absolute(addrs.RootModuleInstance),
 					ChangeSrc: plans.ChangeSrc{
 						Action: plans.NoOp,
-						Before: mustNewDynamicValueStr("baz value"),
-						After:  mustNewDynamicValueStr("baz value"),
+						Before: mustDynamicOutputValue("baz value"),
+						After:  mustDynamicOutputValue("baz value"),
 					},
 					Sensitive: false,
 				},
@@ -43,8 +43,8 @@ func TestTFPlanRoundTrip(t *testing.T) {
 					Addr: addrs.OutputValue{Name: "secret"}.Absolute(addrs.RootModuleInstance),
 					ChangeSrc: plans.ChangeSrc{
 						Action: plans.Update,
-						Before: mustNewDynamicValueStr("old secret value"),
-						After:  mustNewDynamicValueStr("new secret value"),
+						Before: mustDynamicOutputValue("old secret value"),
+						After:  mustDynamicOutputValue("new secret value"),
 					},
 					Sensitive: true,
 				},
@@ -143,6 +143,14 @@ func TestTFPlanRoundTrip(t *testing.T) {
 	}
 }
 
+func mustDynamicOutputValue(val string) plans.DynamicValue {
+	ret, err := plans.NewDynamicValue(cty.StringVal(val), cty.DynamicPseudoType)
+	if err != nil {
+		panic(err)
+	}
+	return ret
+}
+
 func mustNewDynamicValue(val cty.Value, ty cty.Type) plans.DynamicValue {
 	ret, err := plans.NewDynamicValue(val, ty)
 	if err != nil {
@@ -158,4 +166,97 @@ func mustNewDynamicValueStr(val string) plans.DynamicValue {
 		panic(err)
 	}
 	return ret
+}
+
+// TestTFPlanRoundTripDestroy ensures that encoding and decoding null values for
+// destroy doesn't leave us with any nil values.
+func TestTFPlanRoundTripDestroy(t *testing.T) {
+	objTy := cty.Object(map[string]cty.Type{
+		"id": cty.String,
+	})
+
+	plan := &plans.Plan{
+		Changes: &plans.Changes{
+			Outputs: []*plans.OutputChangeSrc{
+				{
+					Addr: addrs.OutputValue{Name: "bar"}.Absolute(addrs.RootModuleInstance),
+					ChangeSrc: plans.ChangeSrc{
+						Action: plans.Delete,
+						Before: mustDynamicOutputValue("output"),
+						After:  mustNewDynamicValue(cty.NullVal(cty.String), cty.String),
+					},
+				},
+			},
+			Resources: []*plans.ResourceInstanceChangeSrc{
+				{
+					Addr: addrs.Resource{
+						Mode: addrs.ManagedResourceMode,
+						Type: "test_thing",
+						Name: "woot",
+					}.Instance(addrs.IntKey(0)).Absolute(addrs.RootModuleInstance),
+					ProviderAddr: addrs.ProviderConfig{
+						Type: "test",
+					}.Absolute(addrs.RootModuleInstance),
+					ChangeSrc: plans.ChangeSrc{
+						Action: plans.Delete,
+						Before: mustNewDynamicValue(cty.ObjectVal(map[string]cty.Value{
+							"id": cty.StringVal("foo-bar-baz"),
+						}), objTy),
+						After: mustNewDynamicValue(cty.NullVal(objTy), objTy),
+					},
+				},
+			},
+		},
+		TargetAddrs: []addrs.Targetable{
+			addrs.Resource{
+				Mode: addrs.ManagedResourceMode,
+				Type: "test_thing",
+				Name: "woot",
+			}.Absolute(addrs.RootModuleInstance),
+		},
+		Backend: plans.Backend{
+			Type: "local",
+			Config: mustNewDynamicValue(
+				cty.ObjectVal(map[string]cty.Value{
+					"foo": cty.StringVal("bar"),
+				}),
+				cty.Object(map[string]cty.Type{
+					"foo": cty.String,
+				}),
+			),
+			Workspace: "default",
+		},
+	}
+
+	var buf bytes.Buffer
+	err := writeTfplan(plan, &buf)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	newPlan, err := readTfplan(&buf)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	for _, rics := range newPlan.Changes.Resources {
+		ric, err := rics.Decode(objTy)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if ric.After == cty.NilVal {
+			t.Fatalf("unexpected nil After value: %#v\n", ric)
+		}
+	}
+	for _, ocs := range newPlan.Changes.Outputs {
+		oc, err := ocs.Decode()
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if oc.After == cty.NilVal {
+			t.Fatalf("unexpected nil After value: %#v\n", ocs)
+		}
+	}
 }

